numba/npyufunc/omppool.cpp

```diff
@@ -107,10 +152,17 @@ parallel_for(void *fn, char **args, size_t *dimensions, size_t *steps, void *dat
         printf("\n");
     }
 
-    #pragma omp parallel
+    // Set the thread mask on the pragma such that the state is scope limited
+    // and passed via a register on the OMP region call site, this limiting
+    // global state and racing
+    #pragma omp parallel num_threads(num_threads), shared(agreed_nthreads)
     {
         size_t * count_space = (size_t *)alloca(sizeof(size_t) * arg_len);
         char ** array_arg_space = (char**)alloca(sizeof(char*) * array_count);
+
+        // tell the active thread team about the number of threads
+        set_num_threads(agreed_nthreads);
```

Comment:
My understanding is that the team is told about the number of threads because the user might write code (or a developer might write it for testing) that gets the number of threads from within a thread. Is this correct?

Reply:
Yes, also that for nested parallism, one of the active thread team will launch the nested region and need to carry the thread mask into the new parallel region.